scripts/linking.py

```diff
@@ -289,37 +297,147 @@ def get_mention_text_and_ids(data: List[data_util.MedMentionExample],
 
     return mention_texts, gold_umls_ids, missing_entity_ids
 
+def get_mention_text_and_ids_by_doc(data: List[data_util.MedMentionExample],
+                                    umls: Dict[str, Any]):
+    """
+    Returns a list of tuples containing a MedMentionExample and the texts and ids contianed in it
 
-def main(medmentions_path: str, umls_path: str, model_path: str, ks: str, thresholds, train: bool = False):
+    Parameters
+    ----------
+    data: List[data_util.MedMentionExample]
+        A list of MedMentionExamples being evaluated
+    umls: Dict[str, Any]
+        A dictionary of UMLS concepts
+    """
+    missing_entity_ids = []  # entities in MedMentions but not in UMLS
 
-    umls_concept_list = load_umls_kb(umls_path)
-    umls_concept_dict_by_id = {c['concept_id']: c for c in umls_concept_list}
+    examples_with_labels = []
 
-    # We need to keep around a map from text to possible canonical ids that they map to.
-    text_to_concept_id: Dict[str, Set[str]] = defaultdict(set)
+    for example in data:
+        mention_texts = []
+        gold_umls_ids = []
+        for entity in example.entities:
+            if entity.umls_id not in umls:
+                missing_entity_ids.append(entity)  # the UMLS release doesn't contan all UMLS concepts
+                continue
 
-    for concept in umls_concept_list:
-        for alias in set(concept["aliases"]).union({concept["canonical_name"]}):
-            text_to_concept_id[alias].add(concept["concept_id"])
+            mention_texts.append(entity.mention_text)
+            gold_umls_ids.append(entity.umls_id)
+            continue
+        examples_with_labels.append((example, mention_texts, gold_umls_ids))
 
-    if train:
-        create_tfidf_ann_index(model_path, text_to_concept_id)
-    ann_concept_aliases_list, tfidf_vectorizer, ann_index = load_tfidf_ann_index(model_path)
+    return examples_with_labels, missing_entity_ids
 
-    candidate_generator = CandidateGenerator(ann_index, tfidf_vectorizer, ann_concept_aliases_list, text_to_concept_id)
-    print('Reading MedMentions ... ')
-    train_examples, dev_examples, test_examples = data_util.read_full_med_mentions(medmentions_path,
-                                                                                   spacy_format=False)
+def eval_spacy_mentions(examples: List[data_util.MedMentionExample],
+                        umls_concept_dict_by_id: Dict[str, Dict],
+                        candidate_generator: CandidateGenerator,
+                        k_list: List[int],
+                        thresholds: List[float],
+                        spacy_model: str):
+    """
+    Evaluates candidate generation using mentions produced by a spacy model. This means that an entity is considered
+    correct if that entity appears anywhere in the abstract
```

Comment:
I started working on the supervised model but then realized that we need the mention-level evaluation (not just the document-level). It would be great to add that as well.

Reply:
With mention-level evaluation, it would be better to merge this function with `eval_gold_mentions` into one to reduce duplicate code. Or instead,  update `eval_gold_mentions` so that it works with gold or predicted mentions.

Reply:
Agree with Iz here, you could do this nicely by making `examples` (doc, MedMentionsExample) tuples, which would provide all the information needed to evaluate either way.